### PR TITLE
Implement ability to pass compiler arguments with `nimble install -- …`

### PR DESCRIPTION
**Pull request by SolitudeSF**
Add `passNim/p` flag to `nimble install` for passing compiler flags. `nimble install --passNim:-d:danger`.

#599

### CODE REVIEW

**Comment by c-blake** [comment]
I think a feature like this would be widely appreciated!

File: src/nimble.nim.cfg [hunk 1]
```diff
@@ -4,3 +4,4 @@
 --path:"./vendor/nim"
 -d:ssl
 -d:nimcore # Enable 'gorge' in Nim's VM. See https://github.com/nim-lang/Nim/issues/8096
+-d:nimOldCaseObjects
```
Comment (genotrance) [inline src/nimble.nim.cfg]: Why is this needed now?

File: src/nimblepkg/options.nim [hunk 2]
```diff
@@ -40,6 +40,7 @@ type
     of actionInstall, actionPath, actionUninstall, actionDevelop:
       packages*: seq[PkgTuple] # Optional only for actionInstall
                                # and actionDevelop.
+      installOptions*: seq[string]
```
Comment (genotrance) [inline src/nimblepkg/options.nim]: `installOptions` isn't too clear a name

Comment (SolitudeSF) [reply src/nimblepkg/options.nim]: well, i need some alternative to `compileOptions2`

File: src/nimble.nim.cfg (hunk 1 above)
Comment (SolitudeSF) [reply src/nimble.nim.cfg]: when i built nimble manually, without koch, it crashed without this flag.

**Comment by genotrance** [comment]
Two general comments:
- Can this be done without the extra `--`? Are there any nimble flags which apply to nim as well that won't get passed through?
- What about `nimble build`?

**Comment by SolitudeSF** [comment]
This can be done without `--`, nimble would just have to sacrifice short `-d` option for deps only.

I was under the impression that nimble build already did that

File: src/nimble.nim.cfg (hunk 1 above)
Comment (genotrance) [reply src/nimble.nim.cfg]: You are using a really old version of nimble. Need to get the latest codebase.

Comment (SolitudeSF) [reply src/nimble.nim.cfg]: Ah, forgot to `pull`

File: src/nimblepkg/options.nim (hunk 2 above)
Comment (SolitudeSF) [reply src/nimblepkg/options.nim]: would `installCompileOptions` be better?

Comment (c-blake) [reply src/nimblepkg/options.nim]: It sure is crystal clear (though a bit of a mouthful, but that's usually the trade-off).

Comment (genotrance) [reply src/nimblepkg/options.nim]: Why not something like `commandLineFlags` or something that suggest they come from the CLI.

File: src/nimblepkg/options.nim [hunk 3]
```diff
@@ -60,7 +61,9 @@ Usage: nimble COMMAND [opts]
 
 Commands:
   install      [pkgname, ...]     Installs a list of packages.
-               [-d, --depsOnly]   Install only dependencies.
+               [-D, --depsOnly]   Install only dependencies.
```
Comment (genotrance) [inline src/nimblepkg/options.nim]: This will surely break someone. Cannot change existing flags.

I presume you are trying to enable `-d:defineToPassNim` but does `-d:xyz` still get caught in the nimble code? Isn't it just looking for `-d`?

File: src/nimblepkg/options.nim (hunk 2 above)
Comment (SolitudeSF) [reply src/nimblepkg/options.nim]: well, i took whats already implemented for `build` command. should it be changed too or just this instance?

File: src/nimblepkg/options.nim (hunk 3 above)
Comment (SolitudeSF) [reply src/nimblepkg/options.nim]: yep, it will. i can add workaround for that.

File: src/nimblepkg/options.nim (hunk 2 above)
Comment (genotrance) [reply src/nimblepkg/options.nim]: I'd prefer that but @dom96 might want to leave it as is and consistent.

File: src/nimblepkg/options.nim (hunk 3 above)
Comment (SolitudeSF) [reply src/nimblepkg/options.nim]: added workaround. now only `-d` is captured by nimble.

**Comment by genotrance** [comment]
Test cases would be great.

Otherwise LGTM, let's wait from @dom96's review.

**Comment by SolitudeSF** [comment]
whats the point of addidng `passNim` to `build` is it has no cli flags by itself? or is this just futureproofing?

**Comment by dom96** [comment]
> or is this just futureproofing?

It is, although I'm willing to make ``nimble build`` an exception just like ``nimble c`` (they already allow ``-d:release`` IIRC).

**Comment by SolitudeSF** [comment]
both `build` and `c` forward all flags to compiler

**Comment by SolitudeSF** [comment]
i dont really understand what to do with custom task. i see the `arguments` and `flags` fields in `Action`, but i dont see it used.

**Comment by dom96** [comment]
> i dont really understand what to do with custom tasks. i see the arguments and flags fields in Action, but i dont see them being used.

Yeah, that's a bit more complex, feel free to leave that for another time

**Comment by SolitudeSF** [comment]
@genotrance, how would i go about testing this? this just uses already existing mechanism that just wasnt utilized fully.

File: src/nimblepkg/options.nim (hunk 2 above)
Comment (dom96) [reply src/nimblepkg/options.nim]: This needs to have ``passNim`` in the name... rename it to ``passNimFlags``.

**Comment by genotrance** [comment]
Well given you have added the ability to pass flags to the install step, you will need to create a test project that needs some -d:xyz to compile and install properly. Perhaps also a custom gc or other Nim flag that results in a special when statement being executed.

**Comment by SolitudeSF** [comment]
Added test. If its correct then that should be it.

**Comment by dom96** [comment]
Awesome, looks good, thank you!
